### PR TITLE
Initial commit of vector database example with new embeddings

### PR DESCRIPTION
**Pull request by colin-openai**
This PR contains a notebook running through an example of using our embeddings to embed Simple Wikipedia and then indexing and searching it in both Weaviate and Pinecone.

### CODE REVIEW

File: examples/vector_databases/Vector_db_introduction.ipynb [hunk 1]
```diff
@@ -0,0 +1,1187 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ad74202e",
+   "metadata": {},
+   "source": [
+    "Thanks for following along, you're now equipped to set up your own vector databases and use embeddings to do all kinds of cool things - enjoy! For more complex use cases please continue to work through the cookbook examples here:\n",
+    "\n",
+    "TODO: Make other cool things to link to"
```
Comment (filipeabperes) [inline examples/vector_databases/Vector_db_introduction.ipynb:1162]: Still TODO I suppose

File: examples/vector_databases/Vector_db_introduction.ipynb [hunk 2]
```diff
@@ -0,0 +1,1187 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "bd99e08e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@retry(wait=wait_random_exponential(min=1, max=20), stop=stop_after_attempt(6))\n",
```
[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb]: Fair point, updated to precede the function that it is referring to

File: examples/vector_databases/Vector_db_introduction.ipynb [hunk 3]
```diff
@@ -0,0 +1,1187 @@
...
+    "    num_tokens = sum(len(article) for article in encoded_corpus)\n",
+    "    cost_to_embed_tokens = num_tokens / 1_000 * 0.0004\n",
+    "    print(\n",
+    "        f\"num_articles={len(encoded_corpus)}, num_tokens={num_tokens}, est_embedding_cost={cost_to_embed_tokens:.2f} USD\"\n",
+    "    )\n",
+    "\n",
+    "    # Embed the corpus\n",
+    "    with concurrent.futures.ThreadPoolExecutor(max_workers=num_workers) as executor:\n",
+    "        futures = [\n",
+    "            executor.submit(get_embeddings, text_batch)\n",
```
[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb]: I've wrapped it in a try/except block so the function will return the exception if get_embeddings fails

File: examples/vector_databases/Vector_db_introduction.ipynb [hunk 4]
```diff
@@ -0,0 +1,1187 @@
...
+   "cell_type": "code",
+   "execution_count": 4,
+   "id": "0c1c73cb",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Found cached dataset wikipedia (/Users/colin.jarvis/.cache/huggingface/datasets/wikipedia/20220301.simple/2.0.0/aa542ed919df55cc5d3347f42dd4521d05ca68751f50dbc32bae2a7f1e167559)\n"
```
[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb]: Fail from me - I've stripped these out and added "ignore warnings" at the top to prevent from reoccurring - I can remove these if necessary

File: examples/vector_databases/Vector_db_introduction.ipynb [hunk 5]
```diff
@@ -0,0 +1,1187 @@
...
+    "        \"\"\" Makes chunks out of an input DataFrame. \"\"\"\n",
+    "        splits = self.splits_num(df.shape[0])\n",
+    "        if splits <= 1:\n",
+    "            yield df\n",
+    "        else:\n",
+    "            for chunk in np.array_split(df, splits):\n",
+    "                yield chunk\n",
+    "    \n",
+    "    def splits_num(self, elements: int) -> int:\n",
+    "        \"\"\" Determines how many chunks DataFrame contians. \"\"\"\n",
```
[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb]: Updated and moved to precede each code chunk

File: examples/vector_databases/Vector_db_introduction.ipynb [hunk 6]
```diff
@@ -0,0 +1,1187 @@
...
+    "        return round(elements / self.batch_size)\n",
```
[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb:423]: This still works if there is a batch larger than batch size, but happy to put ceil in there as well for a cleaner solution if needed

File: examples/vector_databases/Vector_db_introduction.ipynb [hunk 7]
```diff
@@ -0,0 +1,1187 @@
...
+     "text": [
+      "Uploading vectors to content namespace..\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Upsert content vectors in content namespace\n",
+    "print(\"Uploading vectors to content namespace..\")\n",
+    "for batch_df in df_batcher(article_df):\n",
+    "    index.upsert(vectors=zip(batch_df.vector_id, batch_df.content_vector), namespace='content')"
```
[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb:482]: I've left out to keep it fairly simple, but I've added notes in three places to let them know it is a good approach to improve upsert performance

File: examples/vector_databases/Vector_db_introduction.ipynb (hunk 1 above)
Comment (colin-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb]: I've stripped out this TODO, will update it with references once we have more material to link to

File: examples/vector_databases/Vector_db_introduction.ipynb [hunk 8]
```diff
@@ -0,0 +1,1031 @@
...
+    "                                            model=MODEL,\n",
+    "                                            )[\"data\"][0]['embedding']\n",
+    "\n",
+    "    # Query namespace passed as parameter using title vector\n",
+    "    query_result = index.query(embedded_query, \n",
+    "                               namespace=namespace, \n",
+    "                               top_k=top_k)\n",
+    "\n",
+    "    # Print query results \n",
+    "    print(f'\\nMost similar results querying {query} in \"{namespace}\" namespace:\\n')\n",
```
Comment (ted-at-openai) [inline examples/vector_databases/Vector_db_introduction.ipynb:578]: would wrap the {query} in quotation marks too

Comment (ted-at-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb:578]: maybe rephrase to 'Most similar results to "{query}" in "{namespace}" namespace:\n'. Current phrasing feels a bit harder to parse with the querying verb thrown in.

File: examples/vector_databases/Vector_db_introduction.ipynb [hunk 9]
```diff
@@ -0,0 +1,1031 @@
...
+    "    df = pd.DataFrame({'id':ids, \n",
+    "                       'score':scores,\n",
+    "                       'title': [titles_mapped[_id] for _id in ids],\n",
+    "                       'content': [content_mapped[_id] for _id in ids],\n",
+    "                       })\n",
+    "    \n",
+    "    counter = 0\n",
+    "    for k,v in df.iterrows():\n",
+    "        counter += 1\n",
+    "        print(f'Result {counter} with a score of {v.score} is {v.title}')\n",
```
Comment (ted-at-openai) [inline examples/vector_databases/Vector_db_introduction.ipynb:594]: Starting each line with Result #1 #2 etc feels a bit low-information density. Could also do something like "Battle of Bannockburn (score = 0.869)" so that the results are immediately readable without needing to read all the way across the line. Your call.

File: examples/vector_databases/Vector_db_introduction.ipynb [hunk 10]
```diff
@@ -0,0 +1,1031 @@
...
+     },
+     "execution_count": 110,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "class_obj = {\n",
+    "    \"class\": \"Article\",\n",
+    "    \"vectorizer\": \"none\", # explicitly tell Weaviate not to vectorize anything, we are providing the vectors ourselves through our BERT model\n",
```
Comment (ted-at-openai) [inline examples/vector_databases/Vector_db_introduction.ipynb:805]: We're not using BERT! :D

File: examples/vector_databases/Vector_db_introduction.ipynb [hunk 11]
```diff
@@ -0,0 +1,1031 @@
...
+    "    counter += 1\n",
+    "    print(f\"{counter}. Title: {article['title']} Certainty: {article['_additional']['certainty']}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ad74202e",
+   "metadata": {},
+   "source": [
+    "Thanks for following along, you're now equipped to set up your own vector databases and use embeddings to do all kinds of cool things - enjoy! For more complex use cases please continue to work through other cookbook examples in this repo"
```
Comment (ted-at-openai) [inline examples/vector_databases/Vector_db_introduction.ipynb:1006]: missing period

**Review by ted-at-openai — COMMENTED** [review]
Looks good! I left a bunch of style comments. Don't feel obligated to change them all if you disagree or have more pressing things to do.

File: examples/vector_databases/Vector_db_introduction.ipynb (hunk 3 above)
[the remark this answers is not in the document]
Comment (filipeabperes) [reply examples/vector_databases/Vector_db_introduction.ipynb]: I mostly wasn't sure if the Exception would be raised, or if the thread would just fail silently. It seems it would be raised when `result()` is called. If that's the case you can just let the Exception propagate naturally if you think that makes the code cleaner (without the need for the try/except)

File: examples/vector_databases/Vector_db_introduction.ipynb (hunk 7 above)
[the remark this answers is not in the document]
Comment (filipeabperes) [reply examples/vector_databases/Vector_db_introduction.ipynb:482]: yep, there is some demo code in pinecone docs as well, if you want to just refer to that (maybe you already did)

File: examples/vector_databases/Vector_db_introduction.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (filipeabperes) [reply examples/vector_databases/Vector_db_introduction.ipynb]: agree with Teds comment that it seems to make more sense decorating `get_embeddings()`, but I might be missing something

[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb]: Fair point, I did on the other as it is calling get_embeddings but you are right, makes more sense this way. Updated to decorate get_embeddings()

File: examples/vector_databases/Vector_db_introduction.ipynb (hunk 3 above)
[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb]: Agree, I've removed the try/except block to make it cleaner

File: examples/vector_databases/Vector_db_introduction.ipynb (hunk 7 above)
[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb]: I've added a direct reference to their docs for parallel insertion in the Pinecone intro cell

File: examples/vector_databases/Vector_db_introduction.ipynb [hunk 12]
```diff
@@ -0,0 +1,1031 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "cb1537e6",
+   "metadata": {},
+   "source": [
+    "# Vector Database Introduction\n",
```
[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb]: Renamed to be more descriptive and also changed the named of the notebook itself

File: examples/vector_databases/Vector_db_introduction.ipynb [hunk 13]
```diff
@@ -0,0 +1,1031 @@
...
+    "- **Pinecone**\n",
+    "    - *Setup*: Here we setup the Python client for Pinecone. For more details go [here](https://docs.pinecone.io/docs/quickstart)\n",
+    "    - *Index Data*: We'll create an index with namespaces for __titles__ and __content__\n",
+    "    - *Search Data*: We'll test out both namespaces with search queries to confirm it works\n",
+    "- **Weaviate**\n",
+    "    - *Setup*: Here we setup the Python client for Weaviate. For more details go [here](https://weaviate.io/developers/weaviate/current/client-libraries/python.html)\n",
+    "    - *Index Data*: We'll create an index with __title__ search vectors in it\n",
+    "    - *Search Data*: We'll run a few searches to confirm it works\n",
+    "\n",
+    "Once you've run through this notebook you should have a basic understanding of how to setup and use vector databases, and can move on to more complex use cases making use of our embeddings"
```
[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb]: Added

File: examples/vector_databases/Vector_db_introduction.ipynb [hunk 14]
```diff
@@ -0,0 +1,1031 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e2b59250",
+   "metadata": {},
+   "source": [
+    "## Setup\n",
+    "\n",
+    "Here we import the required libraries and set the embedding model that we'd like to use"
```
[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb]: Added

[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb]: I've gone through all of these and tried to simplify - further feedback welcome if still too wordy

File: examples/vector_databases/Vector_db_introduction.ipynb [hunk 15]
```diff
@@ -0,0 +1,1031 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "63b28543",
+   "metadata": {},
+   "source": [
+    "### Create Index\n",
+    "\n",
+    "First we need to create an index, which we'll call `wikipedia-articles`. Once we have an index, we can create multiple namespaces, which can make a single index searchable for various use cases. For more details, consult [this article](https://docs.pinecone.io/docs/namespaces#:~:text=Pinecone%20allows%20you%20to%20partition,different%20subsets%20of%20your%20index.)."
```
[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb]: Updated and done the same for Weaviate

File: examples/vector_databases/Vector_db_introduction.ipynb (hunk 8 above)
Comment (colin-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb]: Updated

File: examples/vector_databases/Vector_db_introduction.ipynb (hunk 9 above)
Comment (colin-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb]: Updated and made consistent across the two vector DBs

File: examples/vector_databases/Vector_db_introduction.ipynb (hunk 10 above)
Comment (colin-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb]: Copy/paste error! Removed and reworded

File: examples/vector_databases/Vector_db_introduction.ipynb (hunk 11 above)
Comment (colin-openai) [reply examples/vector_databases/Vector_db_introduction.ipynb]: Added

**Comment by kacperlukawski** [comment]
Hi, would you mind adding Qdrant as another option? I can provide a working example similar to the created ones.

**Comment by ted-at-openai** [comment]
Couple more suggestions:

- Add installation instructions for pinecone, weaviate, and qdrant_client, as most people won't have them. I think it's especially valuable here since the package names and import names aren't the same. Looks like pinecone-client and weaviate-client and qdrant-client, vs pinecone, weaviate, and qdrant_client.
- Add qdrant to the table of contents/outline at the top
- After running `pip install --upgrade pinecone-client` I immediately run into an error when importing it. Not sure why, but I want to figure it out before we merge

**Comment by ted-at-openai** [comment]
^on the error I'm hitting, I've emailed pinecone support and will wait to hear back.

File: examples/vector_databases/Using_vector_databases_for_embeddings_search.ipynb [hunk 16]
```diff
@@ -0,0 +1,1419 @@
...
+    "### Demo Flow\n",
+    "The demo flow is:\n",
+    "- **Setup**: Import packages and set any required variables\n",
+    "- **Load data**: Load a dataset and embed it using OpenAI embeddings\n",
+    "- **Pinecone**\n",
+    "    - *Setup*: Here we setup the Python client for Pinecone. For more details go [here](https://docs.pinecone.io/docs/quickstart)\n",
+    "    - *Index Data*: We'll create an index with namespaces for __titles__ and __content__\n",
+    "    - *Search Data*: We'll test out both namespaces with search queries to confirm it works\n",
+    "- **Weaviate**\n",
+    "    - *Setup*: Here we setup the Python client for Weaviate. For more details go [here](https://weaviate.io/developers/weaviate/current/client-libraries/python.html)\n",
```
Comment (ted-at-openai) [inline examples/vector_databases/Using_vector_databases_for_embeddings_search.ipynb:30]: "setup" is the noun, "set up" is the verb. These should say "set up".

File: examples/vector_databases/Using_vector_databases_for_embeddings_search.ipynb [hunk 17]
```diff
@@ -0,0 +1,1419 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d939342f",
+   "metadata": {},
+   "source": [
+    "## Weaviate\n",
+    "\n",
+    "The other vector database option we'll explore here is **Weaviate**, which offers both a managed, SaaS option like Pinecone, as well as a self-hosted option. As we've already looked at a cloud vector database, we'll try the self-hosted option here.\n",
```
Comment (ted-at-openai) [inline examples/vector_databases/Using_vector_databases_for_embeddings_search.ipynb:714]: One thing I noticed is that some of the notebook is in future tense (e.g., "we will do X"), some is in present tense (e.g., "we set up Y"), and some is in past tense ("we used Docker"). Worth doing another pass to make them all consistent.

**Comment by ted-at-openai** [comment]
One last suggestion: I think it would be helpful if we precomputed the embeddings, stored them on our CDN, and let people download them, so that they don't have to pay $5 each time they run the example. This is what we've done in some of the other examples. Feel free to pick any file format you like.

Then, if everything runs on your end, we can merge. Thanks again for all the work on this!

File: examples/vector_databases/Using_vector_databases_for_embeddings_search.ipynb (hunk 16 above)
Comment (colin-openai) [reply examples/vector_databases/Using_vector_databases_for_embeddings_search.ipynb:30]: Updated

File: examples/vector_databases/Using_vector_databases_for_embeddings_search.ipynb (hunk 17 above)
Comment (colin-openai) [reply examples/vector_databases/Using_vector_databases_for_embeddings_search.ipynb:714]: Updated to align with future tense

**Comment by colin-openai** [comment]
> One last suggestion: I think it would be helpful if you precomputed the embeddings, stored them on our CDN, and let people download them, so that they don't have to pay $5 each time they run the example. This is what we've done in some of the other examples. Feel free to pick any file format you like. DM me to discuss how to upload to our CDN and what URL we'll want to give it.
> 
> Then, if everything runs on your end, we can merge. Thanks again for all the work on this!

@ted-at-openai these are now resolved
